singularity-eos/closure/mixed_cell_models.hpp

```diff
@@ -97,64 +101,132 @@ bool check_nans(Real const *const a, const std::size_t n, const bool verbose = f
 
 PORTABLE_INLINE_FUNCTION
 bool solve_Ax_b_wscr(const std::size_t n, Real *a, Real *b, Real *scr) {
+  // Simple Jacobi preconditioner
+  //
+  // JMM: This DOES seem to reliably reduce the condition number of
+  // our matrices, suggesting they can be reliably transformed into a
+  // diagonally dominant form. However, it's unclear to me if it
+  // actually helps us solve difficult mixed cells. It seems to me
+  // that when the matrix is ill-conditioned there is some other
+  // feature of the problem making the solve difficult, perhaps the
+  // shape of the residual space, or some source of catastrophic
+  // cancellation we haven't been able to eliminate. That said, it's
+  // basically free, so I decided to keep it.
+  for (std::size_t row = 0; row < n; ++row) {
+    Real maxabs = 0;
+    for (std::size_t column = 0; column < n; ++column) {
+      maxabs = std::max(std::abs(a[row * n + column]), maxabs);
+    }
+    if (maxabs == 0) {
+      // row is all zeros. Matrix is not invertible.
+      return false;
+    }
+    for (std::size_t column = 0; column < n; ++column) {
+      a[row * n + column] /= maxabs;
+    }
+    b[row] /= maxabs;
+  }
+
+  if (n == 1) {                       // JMM: Just in case
+    b[0] = robust::ratio(b[0], a[0]); // re-use b as x
+    // JMM: Special case Cramer's rule for 2x2 matrices, where it is
+    // both more efficient and more accurate. Note that at 3x3 and
+    // greater, it becomes worse in both efficiency and
+    // accuracy.
```

Comment:
Maybe it's just github, but should this comment be under the `else if` block below?

Reply:
oops yep